compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,387 @@
+
+  for (auto i = 0; i < _shape.elements.size(); i++)
+  {
+    const auto &next_tags = _shape.elements.at(i).tags;
+
+    if (next_tags.size() == 0)
+      continue;
+
+    if (is_incremental_tags(next_tags) == false)
+      return false;
+
+    std::string current_tag_str = std::string(std::begin(current_tags), std::end(current_tags));
+    std::string next_tag_str = std::string(std::begin(next_tags), std::end(next_tags));
+
+    if (current_tag_str <= next_tag_str)
+      current_tags = next_tags;
+    else
+      return false;
+  }
+  return true;
+}
+
+luci::CircleReshape *PatternSimulator::createAlternative(loco::Graph *graph)
+{
+  std::string pattern_name =
+    _front_transpose->name() + "/" + _mid_reshape->name() + "/" + _back_transpose->name();
+
+  auto shape_node = graph->nodes()->create<luci::CircleConst>();
+  {
+    shape_node->dtype(loco::DataType::S32);
+    shape_node->size<loco::DataType::S32>(_shape.elements.size());
+    shape_node->rank(_shape.elements.size());
+    for (uint32_t i = 0; i < _shape.elements.size(); i++)
+    {
+      shape_node->at<loco::DataType::S32>(i) = _shape.elements[i].value;
+    }
+    shape_node->shape_status(luci::ShapeStatus::VALID);
+    shape_node->name(pattern_name + "/shape");
+  }
+
+  auto reshape_node = graph->nodes()->create<luci::CircleReshape>();
+  {
+    reshape_node->name(pattern_name);
+    reshape_node->tensor(_front_transpose->a());
+    reshape_node->shape(shape_node);
+    luci::add_origin(reshape_node, luci::get_origin(_mid_reshape));
+  }
+
+  return reshape_node;
+}
+
+bool remove_no_effect_reshape(luci::CircleTranspose *node)
+{
+  // find 'front_transpose - mid_reshape - back_transpose' pattern
+  const auto back_transpose = node;
+  const auto mid_reshape = dynamic_cast<luci::CircleReshape *>(back_transpose->a());
+  {
+    if (not mid_reshape)
+      return false;
+
+    const auto shape_node = dynamic_cast<luci::CircleConst *>(mid_reshape->shape());
+    if (shape_node == nullptr)
+      return false;
+
+    if (loco::succs(mid_reshape).size() != 1)
+      return false;
+  }
+  const auto front_transpose = dynamic_cast<luci::CircleTranspose *>(mid_reshape->tensor());
+  {
+    if (not front_transpose)
+      return false;
+
+    if (loco::succs(front_transpose).size() != 1)
+      return false;
+  }
+
+  // simulate
+  PatternSimulator sim(front_transpose, mid_reshape, back_transpose);
+
+  sim.simulate();
+
+  if (sim.areTransposesUnnecessary() == false)
+    return false;
+
+  // replace
+  auto new_reshape_node = sim.createAlternative(node->graph());
+  replace(node).with(new_reshape_node);
+  return true;
+}
+
+} // namespace
+
+namespace luci
+{
+
+bool RemoveUnnecessaryTransposeNetPass::run(loco::Graph *g)
+{
+  bool changed = false;
+
+  for (auto node : loco::active_nodes(loco::output_nodes(g)))
+  {
+    if (auto transpose_node = dynamic_cast<luci::CircleTranspose *>(node))
+    {
+      if (remove_no_effect_reshape(transpose_node))
```

Comment:
logic is quite long ... TLDR.
1/ please add a diagram like other pass.
2/ please a some comment explaining the logic so that others and future you can understand without difficulties.

Reply:
Thank you for reviewing :smile: 

I catch your point. 
Definity, The code is a bit long, and it is hard to catch the overall things. 
I'll try to add some comments to explain the overall logic.